lib/services/northBound/contextServer.js

```diff
@@ -482,6 +704,130 @@ function handleQuery(req, res, next) {
+    }
+
+    function completeAttributes(attributes, device, callback) {
+        if (attributes && attributes.length !== 0) {
+            logger.debug(context, 'Handling received set of attributes: %j', attributes);
+            callback(null, attributes);
+        } else if (device.lazy) {
+            logger.debug(context, 'Handling stored set of attributes: %j', attributes);
+            var results = device.lazy.map(getName);
+            callback(null, results);
+        } else {
+            logger.debug(context, 'Couldn\'t find any attributes. Handling with null reference');
+            callback(null, null);
+        }
+    }
+
+    function createQueryRequest(attributes, contextEntity, callback) {
+        var actualHandler;
+
+        if (queryHandler) {
+            actualHandler = queryHandler;
+        } else {
+            actualHandler = defaultQueryHandlerNgsi2;
+        }
+
+        async.waterfall([
+            apply(
+                deviceService.getDeviceByName,
+                contextEntity.id,
+                req.headers['fiware-service'],
+                req.headers['fiware-servicepath']),
+        ], function handleFindDevice(error, innerDevice) {
+            if (!innerDevice) {
+                callback(new errors.DeviceNotFound(contextEntity.id));
+            }
+
+            if(!contextEntity.type) {
+                contextEntity.type = innerDevice.type;
+            }
+
+            deviceService.findConfigurationGroup(innerDevice, function(error, device) {
+                var executeCompleteAttributes = apply(
+                    completeAttributes,
+                    attributes,
+                    device
+                    ),
+                    executeQueryHandler = apply(
+                        actualHandler,
+                        contextEntity.id,
+                        contextEntity.type,
+                        req.headers['fiware-service'],
+                        req.headers['fiware-servicepath']
+                    ),
+                    executeAddStaticAttributes = apply(
+                        addStaticAttributes,
+                        attributes,
+                        device
+                    );
+
+                async.waterfall([
+                    executeCompleteAttributes,
+                    executeQueryHandler,
+                    executeAddStaticAttributes
+                ], callback);
+            });
+        });
+
+    }
+
+    function handleQueryContextRequests(error, result) {
+        if (error) {
+            logger.debug(context, 'There was an error handling the query: %s.', error);
+            next(error);
+        } else {
+            logger.debug(context, 'Query from [%s] handled successfully.', req.get('host'));
+            res.status(200).json(result);
+        }
+    }
+
+    logger.debug(context, 'Handling query from [%s]', req.get('host'));
+    var contextEntity = {};
+    contextEntity.id = req.query.id;
```

Comment:
Note that the forward query could be without id, type and/or attrs query parameters (they are optionals). How the code behaves in this situation? I mean:

* When id query param is not present
* When type query param is not present
* When attrs query param is not present

(Probably the answer is in the logic of the inner functions in handleQueryNgsi2() but it is a bit complex so I prefer to ask directly :)

Reply:
Also to make sure that each case is properly covered by unit test.

Reply:
Some tests and features have been added with bbf52286e1edc7551ac554c59fc150c69f273f93 and e7a412efc83892848ec05b12a14065de601f9060

Reply:
Seems to be a coverage good enough. Thanks for the precise references!

NTC